multiuser/machine-auth/che-multiuser-machine-authentication/src/main/java/org/eclipse/che/multiuser/machine/authentication/server/signature/SignatureKeyManager.java

```diff
@@ -108,21 +87,34 @@ public void onEvent(WorkspaceStatusEvent event) {
         };
   }
 
-  /** Returns cached instance of {@link KeyPair} or null when failed to load key pair. */
-  @Nullable
-  public KeyPair getKeyPair(String workspaceId) throws ServerException {
+  /**
+   * Returns instance of {@link KeyPair} for given workspace.
+   *
+   * @throws SignatureKeyManagerException when stored keypair is incorrect (e.g. has bad algorithm
+   *     or keyspec) or other error
+   */
+  public KeyPair getOrCreateKeyPair(String workspaceId) throws SignatureKeyManagerException {
+    SignatureKeyPair keyPair;
     try {
-      return cachedPair.get(workspaceId);
-    } catch (ExecutionException e) {
-      throw new ServerException(e.getCause());
+      try {
+        keyPair = signatureKeyDao.get(workspaceId);
+      } catch (NotFoundException e) {
+        keyPair = generateKeyPair(workspaceId);
+      }
+    } catch (NoSuchAlgorithmException | ServerException | ConflictException ex) {
+      LOG.error(
```

Comment:
Do you think we should log error when somebody tries to generate key pair for non-existing workspace? Maybe lets remove it here and just propagate original exception to a client?

Reply:
Yes i would prefer to leave this log until we fully solve those unclear situations with removed workspaces.